### ISSUE TITLE
DataTransfer.files not available in drag events (using file protocol)

### ISSUE THREAD

**Issue by inukshuk** [electron/electron#9840]
* Electron version: 1.7.3
* Operating system: Linux and macOS

### Expected behavior

When using the file protocol `event.dataTransfer.files` was available to drag events even before the drop event (at least up to Electron 1.6.11) (see related discussion #8720) -- in Electron 1.7.3 the file list is always empty.

### Actual behavior

Files are not available.

### How to reproduce

Register `dragover` or `dragenter` and drag files over the application.

**Comment by inukshuk** [comment]
I have added a simple [test case](https://github.com/inukshuk/electron-quick-start/tree/issue/file-drag) to illustrate the issue. The test script itself is just:

```Javascript
  function report(type) {
    const element = document.createElement('pre')

    document.body.appendChild(element)
    document.addEventListener(type, event => {
      event.preventDefault()

      let num = 0
      let paths = []

      for (let file of event.dataTransfer.files) {
        paths.push(file.path)
        num++
      }

      element.innerHTML =
        `${type} ${num} File(s):\n  ${paths.join('\n  ')}\n`
    })
  }

  report('dragenter')
  report('dragover')
  report('drop')
```

With 1.6.11 the files (and paths) can be accessed by all three event handlers; on 1.7.x only the drop handler gets access to them.

Example screenshot when dragging and dropping 4 files in 1.6.11:
![screenshot from 2017-07-20 16-03-55](https://user-images.githubusercontent.com/325102/28422521-8ecde176-6d68-11e7-92e0-3733bf6165ca.png)

Example screenshot when dragging and dropping the same 4 files in 1.7.5:
![screenshot from 2017-07-20 16-05-34](https://user-images.githubusercontent.com/325102/28422556-aa31683e-6d68-11e7-80a5-8c263c7e74e3.png)

Losing access to the files on drag enter/over is unfortunate, because it means applications cannot implement guidance for users or pre-process files dragged into the application before they are actually dropped (simple example: visually show whether or not it is possible to drop a given file).

**Comment by bpasero** [comment]
We are also seeing this in VS Code. Is this a regression of Chrome or Electron?

**Comment by zcbenz** [comment]
@bpasero This seems to be a regression of Chrome browser, you can verify it by running the code in latest Chrome browser.

**Comment by bpasero** [comment]
@zcbenz ok thanks for letting me know. While this could be a regression, this could also be intentional (i.e. do not allow to sneak into the data transfer unless the user dropped). I will try to file a bug report against Chrome to see the answer.

Anyway, given Electron is around desktop integration I think Electron should lift this protection (or provide an option for it).

**Comment by Tyriar** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=104681&can=1&q=component%3ABlink%3EDataTransfer%20drag%20files&sort=-modified&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified

**Comment by Tyriar** [comment]
See "Protected mode" against the events in the spec https://html.spec.whatwg.org/multipage/dnd.html#dndevents

> Protected mode
>
> For all other events. The formats and kinds in the drag data store list of items representing dragged data can be enumerated, but the data itself is unavailable and no new data can be added.

The kind is either string or file, it doesn't differentiate folder and file is the issue.

**Comment by Tyriar** [comment]
We do get `DataTransferItem.type` which gives the mime type. It doesn't appear to be filled for folders as they don't really have a standard mime type though.

```
gulpfile.js -> 'text/plain'
some.dll    -> 'application/x-msdownload'
somefolder  -> ''
```

**Comment by bpasero** [comment]
@Tyriar yeah it looks like that is the fix in Chrome that causes this behaviour. Your trick with the mime type is unfortunately not working for all cases. I just tried it out with a file that had an unknown extension and it resulted in the type being an empty string too.

**Comment by zcbenz** [comment]
@bpasero If we can find the commit in Chrome that did the change, I would be happy to revert it in Electron.

**Comment by bpasero** [comment]
@zcbenz looks like this is the diff: https://chromium.googlesource.com/chromium/src.git/+/2cb1858d033ef1154182ca529e16703693ae9cf6%5E%21/#F5 (via https://chromium.googlesource.com/chromium/src.git/+/2cb1858d033ef1154182ca529e16703693ae9cf6)

**Comment by Tyriar** [comment]
It's a change to Blink, does Electron do this anywhere yet?

**Comment by bpasero** [comment]
@Tyriar I believe so, when I look at https://github.com/electron/libchromiumcontent/tree/8915338e8cca8679e884efcd6aa5c046b1de57a4/patches (but maybe @zcbenz can confirm)

**Comment by Tyriar** [comment]
Yep looks like it 😄  https://github.com/electron/libchromiumcontent/blob/8915338e8cca8679e884efcd6aa5c046b1de57a4/patches/blink_world_context.patch

**Comment by bpasero** [comment]
Btw VS Code is no longer blocked on this issue, we added a workaround.

**Comment by stevekinney** [comment]
Hey friends—it looks like this is still an issue in 1.7.11. Does anyone have any insight into whether or not this will be fixed in an upcoming version of Electron and/or should I start looking into work arounds? Thanks!

**Comment by OualidYousfi** [comment]
Using 1.8.4 and still an issue for me. DataTransfer.files are not available on onDragEnter event. Nor on onDragOver or drop.

**Comment by lostfictions** [comment]
Can anyone confirm whether 2.0.0+ is still affected by this regression? We'd probably consider upgrading if it would fix this.

**Comment by inukshuk** [comment]
This is still open (you can access `DataTransfer.files` on drag in anything above Electron 1.6).

**Comment by greggman** [comment]
yes it's still an issue

For me I just need to know the types of the files, not their paths. I still get `event.dataTransferItems.types` but it only handles a few types. Examples, when I drag the follow files I see the corresponding type

```
foo.exe    application/x-msdowload
foo.txt    text/plain
foo.json   <empty string>
foo.xlsx   <empty string>
foo.docx   <empty string>
foo.png    image/png
```

If I could register more extention->type settings via some API that would be enough for me although I'm all for getting the actual filenames in Electron

@bpasero is there a link to your workaround?

**Comment by bpasero** [comment]
@greggman we do not really have a fix for this, our workaround was to not rely on this information anymore, but the functionality degraded because of that decision.

**Comment by spenoir** [comment]
I can confirm that in Electron 2.0.5 drag Over events get empty strings from DataTransfer object, however Drop events get the correct values. The types get set fine just empty string values on drag Over events at least

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in Electron v4.2.x or later, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons. This issue will be closed 7 days from now if there is no response.

Thanks in advance! Your help is appreciated.

**Comment by inukshuk** [comment]
This is still open in all the latest Electron releases.

I've been using `event.dataTransferItems.types` to check the mime-types as @greggman suggests above. Being able to check for types mitigates the issue and, I would imagine, should be sufficient to handle drag events in most cases (with more involved checks happening on drop when you can see the actual files).

**Comment by astoilkov** [comment]
@inukshuk Using the mime-types is actually not sufficient because sometimes they are empty (at least on macOS).

**Comment by inukshuk** [comment]
@astoilkov yes, it's only a workaround. It will only work if the mime-types in question are properly registered at the OS level (I think! I've only used well-known types so far, it's possible that this is registered somewhere in Chromium). Also, it does not give you a clear mapping to files; but having the types available at least mitigates the issue in some cases. I was just pointing this out, because the issue was about the be closed and I think we'll have to decide, in principle, if this is something Electron wants to fix.

I don't think there's a chance that Chromium would decide to go back to the old behaviour of including the files for the `file://` protocol so the only way to fix this is to backport the old behaviour and maintain it going forward. This would definitely be the best solution for Electron users, but I'm not it's worth the effort. (For the record, I've had this on my TODO list for two years, but since the workaround using mime-types is good enough for our app so far, I haven't looked into it.)

**Comment by ggreco** [comment]
This is still open in electron 8 and electron 10.

@inukshuk workaround seems to not work anymore, since DataTransferItems is not present at all in the dragover event.

**Comment by inukshuk** [comment]
This would be very surprising and it's still working for me in electron 9. How did you try to access the items? The list should be at `dataTransfer.items` on the drag event object.

**Comment by GiancarlosIO** [comment]
😭

**Comment by DWboutin** [comment]
I'm on Electron 11 and I'm not able to get the items or files from `event.dataTransfer` too

**Comment by berkon** [comment]
Any update on this? Getting this functionality back would help a lot to make software more user friendly. My scenario is:

A user can drag/drop an image into a rectangle. If the image is very small it shall be placed into a corner. Therefore I'm highlighting the corner were the mouse hovers to show the user where it will be dropped. Now if the image is bigger, it shall fully fill the rectangle. In that case I would like to highlight the whole rectangle (and not the corners anymore). In order to achieve this, I must obtain the width/height of the image. And for this I need the path of the image. But as mentioned in the posts above, I'm only getting the path when a drop occours. Thus I cannot decide how to highlight the rectangle correctly.

**Comment by dceddia** [comment]
I have a fix for this working locally. It more or less reverts the [diff mentioned earlier](https://chromium.googlesource.com/chromium/src.git/+/2cb1858d033ef1154182ca529e16703693ae9cf6%5E%21/#F5) in Chromium/Blink. It sets the DataTransferAccessPolicy to the more permissive "Readable" (instead of "TypesReadable") when the document is from a file:// or localhost URL.

I'm not sure if there's appetite for a Blink patch? But I can try to make a PR if it's viable. (this would also be my first Electron PR and I might need some guidance)

**Comment by nornagon** [comment]
@dceddia I don't really like the idea of diverging from Chromium here. Perhaps it might be possible to achieve the desired result by exposing some API in the main process that would allow fetching information about drags in progress from a webContents, and then using IPC to coordinate between the renderer and the main process? That would be more flexible, in that it would be possible for the main process to securely apply a policy for when this data should be accessible, instead of "only on file:// or localhost", which is a somewhat restrictive policy, and which the patch you linked doesn't allow for. Further, apps that didn't explicitly opt into the changed behavior would not be affected.

**Comment by dceddia** [comment]
@nornagon That seems like a cleaner way to go. Maybe the renderer could pass one of the opaque `DragEvent`s to main via IPC, and main could pull the underlying file info, via some API on webContents? I don't know much about Chromium/Electron's internals so I'm not sure where to start with this, but it seems like it should be possible.

**Comment by glocore** [comment]
Bump! I'd love to have this work. Similar to @berkon's [comment](https://github.com/electron/electron/issues/9840#issuecomment-900558314), I'd like to be able to show different UI based on the type of file being dragged into the window.
